python/ray/experimental/data_loader/dataset.py

```diff
@@ -0,0 +1,202 @@
+        num_reducers (optional, int): The number of shuffler reducers. Default
+            is the number of trainers x the number of cores on the master
+            (rank 0) worker.
+        max_concurrent_epochs (optional, int): The maximum number of epochs
+            whose shuffling stages should execute concurrently. Default is 2.
+    """
+    def __init__(
+            self,
+            filenames: List[str],
+            num_epochs: int,
+            num_trainers: int,
+            batch_size: int,
+            rank: int,
+            num_rounds: int = 2,
+            num_reducers: int = None,
+            max_concurrent_epochs: int = 2,
+            max_batch_queue_size: int = 100):
+        if num_reducers is None:
+            num_reducers = num_trainers * get_num_cpus()
+
+        self._batch_size = batch_size
+
+        # TODO(Clark): Find way to do this without batch consumer proxy queue.
+        if rank == 0:
+            # rank == 0 --> master process
+            # Create the batch queue. Trainers will consume GPU batches
+            # through this batch queue.
+            # TODO(Clark): If the shuffle API was changed to take num_trainers
+            # batch consumers instead of a single batch consumer, instead of
+            # having a single batch multi-queue, we could create num_trainers
+            # batch queues, have each batch consumer close over their
+            # corresponding batch queue, and have each trainer reference their
+            # single batch queue.
+            self._batch_queue = MultiQueue(
+                num_trainers, max_batch_queue_size,
+                name=MULTIQUEUE_ACTOR_NAME, connect=False)
+            # Kick off shuffle.
+            # TODO(Clark): Move the shuffle kickoff to an init() method so the
+            # user can better control when the shuffling starts?
+            self._shuffle_result = ray.remote(shuffle_from_memory).remote(
+                filenames,
+                functools.partial(
+                    batch_consumer, self._batch_queue, batch_size),
+                num_epochs,
+                num_rounds,
+                num_reducers,
+                num_trainers,
+                max_concurrent_epochs,
+                collect_stats=False)
+        else:
+            # rank != 0 --> worker process
+            # Connect to the batch queue.
+            self._batch_queue = MultiQueue(
+                num_trainers, max_batch_queue_size,
+                name=MULTIQUEUE_ACTOR_NAME, connect=True)
+            self._shuffle_result = None
+
+        self._rank = rank
+
+    def __iter__(self):
+        """
+        This iterator yields GPU batches from the shuffling queue.
+        """
+        leftover_batches = None
+        while True:
+            # TODO(Clark): Add get_up_to queue method that can fetch up to
+            # some number of batches in a single RPC.
+            batches = self._batch_queue.get(self._rank, block=True)
```

Comment:
Consider using a ring buffer-like approach (with logical concatenation):

Order of batches: [xxxxx] [yyy] [zz]
1. buffer: [xxxxx] -> yield [xx], [xx]
2. buffer: [x][yyy] -> yield [xy], [yy]
3. buffer: [zz] -> yield [zz]